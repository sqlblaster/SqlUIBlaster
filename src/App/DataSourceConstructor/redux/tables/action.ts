import { Action } from 'src/components/query-builder/store/models/action-types';
import { Table } from '../../schemas';

export type TablesAction = Action<'SET_TABLES', { tables: Table[] }>;

export const setTables = (tables: Table[]): TablesAction => ({
  type: 'SET_TABLES',
  tables
});
