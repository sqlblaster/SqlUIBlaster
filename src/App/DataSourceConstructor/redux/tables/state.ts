import { Table } from '../../schemas';

export class TablesState {
  public tables: Table[] = [];
}
