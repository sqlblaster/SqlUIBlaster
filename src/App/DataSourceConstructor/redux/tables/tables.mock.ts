import { Column, Table } from '../../schemas';

export const tablesMock = [
  new Table({
    name: 'Orders',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'Details',
        type: 'string',
        enums: ['Detail1', 'Detail2']
      }),
      new Column({
        name: 'Created At',
        type: 'Date'
      }),
      new Column({
        name: 'Quantity',
        type: 'number'
      }),
      new Column({
        name: 'Paid',
        type: 'boolean'
      }),
      new Column({
        name: 'ProductID',
        type: 'string',
        foreignTableName: 'Products'
      }),
      new Column({
        name: 'UserID',
        type: 'string',
        foreignTableName: 'Users'
      })
    ]
  }),
  new Table({
    name: 'Users',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'AddressID',
        type: 'number',
        foreignTableName: 'Addresses'
      }),
      new Column({
        name: 'Username',
        type: 'string'
      }),
      new Column({
        name: 'AdminID',
        type: 'string',
        foreignTableName: 'Users'
      })
    ]
  }),
  new Table({
    name: 'Products',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'Product name',
        type: 'string'
      }),
      new Column({
        name: 'Available amount',
        type: 'number'
      }),
      new Column({
        name: 'AddedBy',
        type: 'string',
        foreignTableName: 'Users'
      }),
      new Column({
        name: 'CurrencyID',
        type: 'string',
        foreignTableName: 'Currency'
      }),
      new Column({
        name: 'LocalCurrencyID',
        type: 'string',
        foreignTableName: 'Currency'
      })
    ]
  }),
  new Table({
    name: 'Addresses',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'StreetID',
        type: 'string',
        foreignTableName: 'Streets'
      })
    ]
  }),
  new Table({
    name: 'Streets',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'Street name',
        type: 'string'
      })
    ]
  }),
  new Table({
    name: 'Currency',
    primaryKeyColumn: 'ID',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'name',
        type: 'string'
      })
    ]
  })
];
