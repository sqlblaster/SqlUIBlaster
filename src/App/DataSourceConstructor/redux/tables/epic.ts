import { ofType } from 'redux-observable';
import { of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { raiseGlobalError } from 'src/store/global-error/action';
import { Epic } from 'src/store/models/Epic';
import { SetDatabaseSchemaAction } from '../database-schema/action';
import { setTables } from './action';

export const extractTablesEpic: Epic<SetDatabaseSchemaAction> = (
  action$,
  state$
) =>
  action$.pipe(
    ofType<SetDatabaseSchemaAction>(
      'SET_DATABASE_SCHEMA'
    ),
    map(({ databaseSchema }) =>
      databaseSchema
        ? setTables(databaseSchema.tables)
        : raiseGlobalError('Couldn\'t fetch database schema')
    ),
    catchError((err) => {
      console.error(err && err.xhr && err.xhr.response);

      return of(raiseGlobalError('Couldn\'t fetch database schema'));
    })
  );
