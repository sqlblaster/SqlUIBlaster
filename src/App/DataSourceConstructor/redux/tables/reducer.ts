import { Reducer } from 'redux';
import { Table } from '../../schemas';
import { TablesAction } from './action';

export const tables: Reducer<Table[], TablesAction> = (
  state = [],
  { type, tables: newTables }
) => {
  return type === 'SET_TABLES' ? newTables : state;
};
