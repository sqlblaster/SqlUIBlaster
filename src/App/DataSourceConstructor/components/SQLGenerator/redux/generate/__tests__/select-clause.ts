import { tablesMock } from 'src/App/DataSourceConstructor/redux/tables/tables.mock';
import { generateSQLQuery, SQLGeneratorStateProps } from '..';

const state: SQLGeneratorStateProps = {
  selectedTable: tablesMock[0],
  customColumns: [],
  filters: [],
  aggregations: [],
  groupings: [],
  orders: [],
  rowLimit: null,
  tables: tablesMock
};

describe('FROM clause', () => {
  it('should render * when there is no aggregations and groupings', () => {
    expect(generateSQLQuery(state)).toMatchSnapshot();
  });

  it.skip('should render groupings when there is only groupings', () => {
    // empty block
  });

  it.skip('should render aggregations when there is only aggregations', () => {
    // empty block
  });

  it.skip(`should render aggregations and groupings
      when there is aggregations and groupings `, () => {
    // empty block
  });
});
