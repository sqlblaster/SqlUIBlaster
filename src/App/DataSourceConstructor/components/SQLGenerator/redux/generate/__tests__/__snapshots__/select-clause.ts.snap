// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FROM clause should render * when there is no aggregations and groupings 1`] = `
"
    SELECT
      *
    FROM \\"orders\\" "
`;

exports[`FROM clause should render groupings when there is only groupings 1`] = `
"
    SELECT
      \\"balance1\\".\\"amount\\" AS \\"amount (balance1)\\",\\"balance1\\".\\"name\\" AS \\"name (balance1)\\"
    FROM \\"orders\\" \\"balance1\\"
    GROUP BY \\"balance1\\".\\"amount\\",\\"balance1\\".\\"name\\"
    ORDER BY \\"balance1\\".\\"amount\\" ASC,\\"balance1\\".\\"name\\" ASC"
`;
